docs/Outputs.md

```diff
@@ -0,0 +1,225 @@
+The `endpoint` function returns a factory function for building a specialized transformation for sending tables to the external service.
+The factory function is used to create multiple ways of sending data to the service, as such it accepts a `mapFn` to map each record in a table to a proper object.
+The fields in that object are used by the endpoint to build the message for the external service.
+Every external service must define how that object must be shaped for the communication to happen.
+
+For example:
+
+```flux
+import ext "external_service"
+
+someFilter = (r) => ...
+someOtherFilter = (r) => ...
+
+// Obtain the factory function with this configuration for the endpoint.
+endpoint = ext.endpoint(url: "https://url/to/external/service")
+
+someMapFn = (r) => {
+   // Generate payload from record.
+   return {title: r.someTag, body: string(v: r._value)}
+})
+someOtherMapFn = (r) => {
+   // Generate payload from record.
+   return {title: r.someOtherTag, body: string(v: r._value)}
+})
+
+// Invoke the factory to get custom transformations.
+to1 = endpoint(mapFn: someMapFn)
+to2 = endpoint(mapFn: someOtherMapFn)
+
+from(...)
+    |> range(...)
+    |> filter(fn: someFilter)
+    |> to1()
+
+from(...)
+    |> range(...)
+    |> filter(fn: someOtherFilter)
+    |> to2()
+```
+
+The output of `to` functions contains the result of the communication with the external service and varies on a per-service base.
+
+
+### An Example: Slack
+
+We provide below an example implementation for a Slack endpoint.
+The package provides an `endpoint` function that return the Slack endpoint.
+It also provides helpers to send a single message in a channel and build the request header.
+
+```flux
+package slack
+
+import "http"
+import "json"
+
+// This can be globally overridden in another package with:
+// slack.defaultURL = "https://slack.com/api/chat.anotherURL"
+option defaultURL = "https://slack.com/api/chat.postMessage"
+
+// Util function to create the header.
+header = (token, ctype="application/json") => ({
+    "Authorization": token,
+    "Content Type": ctype
+})
+
+// `message` sends a single message to a Slack channel.
+message = (url=defaultURL, header, channel, text) => {
+    data = {
+        channel: channel,
+        text: text
+    }
+    enc = json.encode(data)
+    return http.post(header: header, url: url, data: enc)
+}
+
+// `endpoint` creates the endpoint for the Slack external service.
+// The returned factory function accepts a `mapFn` parameter.
+// The `mapFn` must return an object with `channel` and `text` fields.
+endpoint = (url=defaultURL, token) =>
+    // The factory function.
+    (mapFn) =>
+        (tables=<-) => tables
+            |> map(fn: mapFn)
+            |> map(fn: (r) => {
+                resp = message(url: url, header: header(token), channel: r.channel, text: r.text)
+                return {r with status: resp.status}
+            })
```

Comment:
Hmm, I really like the map patern but it means that the data returned from the endpoint is not a pass through rather its the result of the mapFn function. We need the original data to be returned with the status.

Perhaps we should do it like this:

```
endpoint = (url=defaultURL, token) =>
    // The factory function.
    (mapFn) =>
        (tables=<-) => tables
            |> map(fn: (r) => {
                data = mapFn(r)
                resp = message(url: url, header: header(token), channel: data.channel, text: data.text)
                return {r with status: resp.status}
            })
```

Reply:
I see your point 👍